Create new account

mikedotexe opened the pull request:
This adds cross-contract calls to a contract that has the [linkdrop contract](https://github.com/near/near-linkdrop).
So essentially when you deploy and initialize the crossword smart contract, you point to an account that's the "creator account." On testnet this would be `testnet` and on mainnet this would be `near`, which both have the linkdrop contract on there. (For testing, I've deployed the linkdrop contract to `linkdrop.demo.testnet` and have been using that.)

Instructions for compiling:

```
cd contract
./build.sh
```

You may want to edit the file `testnet.sh` in there to point to your custom contract.

I personally like to run this on a custom account (as deployed to one create via `near dev-deploy…`) and so when running the frontend you would type:

`env CONTRACT_NAME=xword.demo.testnet npm run start`

and use whatever your account is instead.

mikedotexe committed:
add base54 to utf8 function to utils

mikedotexe committed:
frontend calls claim_reward_new_account and claim_reward, and grab the linkdrop account from the state

mikedotexe committed:
add google analytics

mikedotexe committed:
update basic CSS

mikedotexe committed:
fix node scripts so "dev" uses the dev-deploy and account

mikedotexe committed:
add claim_reward_new_account and use callbacks for transfer and creation of new account

mikedotexe committed:
update testnet script to include initialization function and args

mikedotexe committed:
update crossword Wasm binary

ChaoticTempest commented on src/utils.js:
```diff
@@ -90,5 +97,6 @@ function parseSolutionSeedPhrase(data, gridData) {
   viewMethodOnContract,
-  parseSolutionSeedPhrase
+  parseSolutionSeedPhrase,
+  b64ToUtf8: b64toUtf8
```
Not too familiar with JS, but this looks weird. I'm not sure if this is fully required, since wouldn't `b64toUtf8` suffice instead of `b64toUtf8: b64toUtf8` for this export?

mikedotexe replied:
yeah good point, I had a capital "T" lowercase "t" thing I didn't notice, thanks!

ChaoticTempest commented on src/App.js:
```diff
@@ -1,18 +1,28 @@
+      setNeedsNewAccount(true);
+    } else {
+      setNeedsNewAccount(false);
+    }
```
Could be shortened to:
```suggestion
    setNeedsNewAccount(e.target.value === 'create-account');
```

ChaoticTempest commented on src/App.js:
```diff
@@ -27,30 +37,89 @@ const App = ({ nearConfig, data }) => {
+          } else {
+            setClaimError("Couldn't transfer reward to that account, please try another account name or create a new one.")
+          }
+        }
```
Could reduce the outer `if` block to the following, since the error message is the only thing dependent on `needsNewAccount`
```suggestion
        // Look for base64-encoded "false"
        if (tx_success_value === 'true') {
          // This tells the React app that it's solved and claimed
          setSolvedPuzzle(false);
          setClaimError('');

          // Clean up and get ready for next puzzle
          localStorage.removeItem('playerSolvedPuzzle');
          localStorage.removeItem('guesses');
        } else {
          if (needsNewAccount) {
            setClaimError('Could not create that account, please try another account name.');
          } else { 
            setClaimError("Couldn't transfer reward to that account, please try another account name or create a new one.");
          }
        }
```

ChaoticTempest commented on contract/src/lib.rs:
```diff
@@ -1,16 +1,70 @@
 mod debugging;
 
+use near_sdk::collections::{LookupMap, UnorderedSet};
+use near_sdk::serde_json::{self};
```
```suggestion
use near_sdk::serde_json;
```

ChaoticTempest commented on contract/src/lib.rs:
```diff
@@ -65,18 +126,38 @@ pub struct Puzzle {
+    creator_account: AccountId,
+}
+
+/// When you want to have a "new" function initialize a smart contract,
+/// you'll likely want to follow this pattern of having a default implementation that panics,
+/// directing the user to call the initialization method. (The one with the #[init] macro)
+impl Default for Crossword {
```
We should probably also mention that `#[derive(PanicOnDefault)]` would do this for us instead of having to manually implement it

mikedotexe replied:
you're right! you're good at code reviews

ChaoticTempest reviewed (changes requested):
just a few things to make the code look cleaner, but looking pretty solid with all the comments/docs. This is shaping up to be a really good zero-to-hero starter project

mikedotexe committed:
Update src/App.js

you smart

Co-authored-by: Phuong Nguyen <ChaoticTempest@users.noreply.github.com>

mikedotexe committed:
Update src/App.js

thanks! looks like a "I'll remember to remove this copy/pasting" and then didn't :)

Co-authored-by: Phuong Nguyen <ChaoticTempest@users.noreply.github.com>

mikedotexe committed:
Update contract/src/lib.rs

Co-authored-by: Phuong Nguyen <ChaoticTempest@users.noreply.github.com>

mikedotexe committed:
add PanicOnDefault, fix import according to PR review

mikedotexe committed:
fix util function name for frontend

mikedotexe committed:
Merge branch 'create-new-account' of github.com:near-examples/near-crossword into create-new-account

# Conflicts:
#	src/App.js

mikedotexe committed:
updated Wasm
